From 3cc230c5bcc0ee42a07b6b155ee3b29fb99c07c1 Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Thu, 11 Dec 2025 16:49:11 +0100
Subject: [PATCH] usb: dwc2: add runtime PM routines for debugfs

Debug file system (regdump) requires the core to exit power saving,
before the registers can be dumped. Not doing so lead to platform
freeze. Debugfs regset32 optionally manages runtime PM get_sync/put
calls when the device struct is provided.

So add runtime suspend and resume routines to dwc2 debugfs driver.
A specific device is allocated, to manage the debugfs read (resume),
requests separately, not to disturb the bus suspended state, as
commented in dwc2_runtime_suspend() routine. The real power saving
entry is managed from the platform dwc2_runtime_suspend() routine,
to ensure all parts are actually suspended (gadget, host, debugfs).

Provide device struct to regset32. Take care to manage concurrency with
bus suspend states by checking check device suspend state.
The power saving routines are widely called from IRQ subroutines, so
take care to manage the critical sections.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: I92fef765058150f052505a7c99506ac6502321ff
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/515564
Reviewed-by: Christian BRUEL <christian.bruel@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/usb/dwc2/core.h      |   3 +
 drivers/usb/dwc2/core_intr.c |  12 ++-
 drivers/usb/dwc2/debugfs.c   | 147 ++++++++++++++++++++++++++++++++++-
 drivers/usb/dwc2/drd.c       |  13 +++-
 drivers/usb/dwc2/gadget.c    |  20 +++--
 drivers/usb/dwc2/hcd.c       |  11 +--
 drivers/usb/dwc2/platform.c  |  57 ++++++++++++++
 7 files changed, 243 insertions(+), 20 deletions(-)

--- a/drivers/usb/dwc2/core.h
+++ b/drivers/usb/dwc2/core.h
@@ -854,6 +854,8 @@ struct dwc2_hregs_backup {
  * These are common for both host and peripheral modes:
  *
  * @dev:                The struct device pointer
+ * @ddev:		Device struct used to manage debugfs access with power
+ *			saving modes (runtime PM).
  * @regs:		Pointer to controller regs
  * @hw_params:          Parameters that were autodetected from the
  *                      hardware registers
@@ -1060,6 +1062,7 @@ struct dwc2_hregs_backup {
  */
 struct dwc2_hsotg {
 	struct device *dev;
+	struct device *ddev;
 	void __iomem *regs;
 	/** Params detected from hardware */
 	struct dwc2_hw_params hw_params;
--- a/drivers/usb/dwc2/core_intr.c
+++ b/drivers/usb/dwc2/core_intr.c
@@ -568,10 +568,20 @@ static void dwc2_handle_usb_suspend_intr
 			if (hsotg->wq_gadget)
 				queue_delayed_work(hsotg->wq_gadget, &hsotg->dw_enumtimeout,
 						   msecs_to_jiffies(2000));
+
+			if (hsotg->rpm_suspended) {
+				dev_dbg(hsotg->dev, "enumeration raced with dwc2_port_suspend()\n");
+				hsotg->rpm_suspended = false;
+				pm_runtime_get(hsotg->dev);
+			}
+
 			return;
 		}
 		if (dsts & DSTS_SUSPSTS) {
-			dwc2_gadget_enter_lp(hsotg);
+			if (!hsotg->rpm_suspended) {
+				hsotg->rpm_suspended = true;
+				pm_runtime_put(hsotg->dev);
+			}
 
 			/*
 			 * Change to L2 (suspend) state before releasing
--- a/drivers/usb/dwc2/debugfs.c
+++ b/drivers/usb/dwc2/debugfs.c
@@ -126,7 +126,11 @@ static const struct file_operations test
 static int state_show(struct seq_file *seq, void *v)
 {
 	struct dwc2_hsotg *hsotg = seq->private;
-	int idx;
+	int idx, ret;
+
+	ret = pm_runtime_resume_and_get(hsotg->ddev);
+	if (ret)
+		return ret;
 
 	seq_printf(seq, "DCFG=0x%08x, DCTL=0x%08x, DSTS=0x%08x\n",
 		   dwc2_readl(hsotg, DCFG),
@@ -168,6 +172,8 @@ static int state_show(struct seq_file *s
 		seq_puts(seq, "\n");
 	}
 
+	pm_runtime_put(hsotg->ddev);
+
 	return 0;
 }
 DEFINE_SHOW_ATTRIBUTE(state);
@@ -185,7 +191,11 @@ static int fifo_show(struct seq_file *se
 	struct dwc2_hsotg *hsotg = seq->private;
 	int fifo_count = dwc2_hsotg_tx_fifo_count(hsotg);
 	u32 val;
-	int idx;
+	int idx, ret;
+
+	ret = pm_runtime_resume_and_get(hsotg->ddev);
+	if (ret)
+		return ret;
 
 	seq_puts(seq, "Non-periodic FIFOs:\n");
 	seq_printf(seq, "RXFIFO: Size %d\n", dwc2_readl(hsotg, GRXFSIZ));
@@ -205,6 +215,8 @@ static int fifo_show(struct seq_file *se
 			   val & FIFOSIZE_STARTADDR_MASK);
 	}
 
+	pm_runtime_put(hsotg->ddev);
+
 	return 0;
 }
 DEFINE_SHOW_ATTRIBUTE(fifo);
@@ -230,6 +242,11 @@ static int ep_show(struct seq_file *seq,
 	int index = ep->index;
 	int show_limit = 15;
 	unsigned long flags;
+	int ret;
+
+	ret = pm_runtime_resume_and_get(hsotg->ddev);
+	if (ret)
+		return ret;
 
 	seq_printf(seq, "Endpoint index %d, named %s,  dir %s:\n",
 		   ep->index, ep->ep.name, decode_direction(ep->dir_in));
@@ -252,6 +269,8 @@ static int ep_show(struct seq_file *seq,
 		   dwc2_readl(hsotg, DIEPTSIZ(index)),
 		   dwc2_readl(hsotg, DOEPTSIZ(index)));
 
+	pm_runtime_put(hsotg->ddev);
+
 	seq_puts(seq, "\n");
 	seq_printf(seq, "mps %d\n", ep->ep.maxpacket);
 	seq_printf(seq, "total_data=%ld\n", ep->total_data);
@@ -770,11 +789,129 @@ static int dr_mode_show(struct seq_file
 }
 DEFINE_SHOW_ATTRIBUTE(dr_mode);
 
+static int dwc2_debugfs_runtime_checks(struct dwc2_hsotg *hsotg)
+{
+	/* RPM suspend has been requested to put the core into low power */
+	if (hsotg->rpm_suspended)
+		return -EBUSY;
+
+	/* Connected device isn't suspended (resumed by bus?) */
+	if (dwc2_is_device_mode(hsotg)) {
+		if (dwc2_is_device_connected(hsotg) && !(dwc2_readl(hsotg, DSTS) & DSTS_SUSPSTS))
+			return -EBUSY;
+	}
+
+	return 0;
+}
+
+static int dwc2_debugfs_runtime_suspend(struct device *dev)
+{
+	struct dwc2_hsotg *hsotg = dev_get_drvdata(dev);
+	unsigned long flags;
+
+	dev_dbg(hsotg->dev, "%s lx_state %d\n", __func__, hsotg->lx_state);
+
+	spin_lock_irqsave(&hsotg->lock, flags);
+	if (!dwc2_debugfs_runtime_checks(hsotg)) {
+		/* Balance the pm_runtime_get() call from the Resume ISR */
+		pm_runtime_put(hsotg->dev);
+		hsotg->rpm_suspended = true;
+
+		/*
+		 * gadget has been resumed from suspended state, due to resume ISR
+		 * put it back to suspend state, as there won't be a suspend IRQ here.
+		 */
+		if (hsotg->suspended_from == USB_STATE_CONFIGURED) {
+			usb_gadget_set_state(&hsotg->gadget, USB_STATE_SUSPENDED);
+			call_gadget(hsotg, suspend);
+		}
+	}
+	spin_unlock_irqrestore(&hsotg->lock, flags);
+
+	return 0;
+}
+
+static int dwc2_debugfs_runtime_resume(struct device *dev)
+{
+	struct dwc2_hsotg *hsotg = dev_get_drvdata(dev);
+	unsigned long flags;
+	int ret = 0;
+
+	spin_lock_irqsave(&hsotg->lock, flags);
+	dev_dbg(hsotg->dev, "%s lx_state %d\n", __func__, hsotg->lx_state);
+
+	/* Exiting power saving mode will raise a resume IRQ */
+	if (hsotg->in_ppd) {
+		ret = dwc2_exit_partial_power_down(hsotg, 0, true);
+		if (ret)
+			dev_err(hsotg->dev, "exit power_down failed %d\n", ret);
+	}
+
+	if (hsotg->params.power_down == DWC2_POWER_DOWN_PARAM_NONE &&
+	    hsotg->bus_suspended && !hsotg->params.no_clock_gating) {
+		if (dwc2_is_device_mode(hsotg)) {
+			dwc2_gadget_exit_clock_gating(hsotg, 0);
+		} else {
+			spin_unlock_irqrestore(&hsotg->lock, flags);
+			dwc2_host_exit_clock_gating(hsotg, 0);
+			spin_lock_irqsave(&hsotg->lock, flags);
+		}
+	}
+
+	spin_unlock_irqrestore(&hsotg->lock, flags);
+
+	return ret;
+}
+
+static const struct dev_pm_ops dwc2_debugfs_pm_ops = {
+	RUNTIME_PM_OPS(dwc2_debugfs_runtime_suspend, dwc2_debugfs_runtime_resume, NULL)
+};
+
+static void dwc2_debugfs_ddev_release(struct device *dev)
+{
+	struct dwc2_hsotg *hsotg = container_of(&dev, struct dwc2_hsotg, ddev);
+
+	kfree(hsotg->ddev);
+	hsotg->ddev = NULL;
+}
+
+static struct device_type dwc2_debugfs_type = {
+	.name =		"dwc2_debugfs",
+	.pm =		&dwc2_debugfs_pm_ops,
+};
+
 int dwc2_debugfs_init(struct dwc2_hsotg *hsotg)
 {
 	int			ret;
 	struct dentry		*root;
 
+	hsotg->ddev = kzalloc(sizeof(*hsotg->ddev), GFP_KERNEL);
+	if (!hsotg->ddev)
+		return -ENOMEM;
+
+	hsotg->ddev->parent = hsotg->dev;
+	hsotg->ddev->type = &dwc2_debugfs_type;
+	hsotg->ddev->release = dwc2_debugfs_ddev_release;
+	ret = dev_set_name(hsotg->ddev, "%s-debugfs", dev_name(hsotg->dev));
+	if (ret)
+		return dev_err_probe(hsotg->dev, ret, "failed to set debugfs dev name\n");
+
+	/*
+	 * separate device to manage power saving modes (by using PM runtime) independently of the
+	 * bus active/suspended sates.
+	 */
+	ret = device_register(hsotg->ddev);
+	if (ret) {
+		/* Free resources in dwc2_debugfs_release */
+		put_device(hsotg->ddev);
+		return dev_err_probe(hsotg->dev, ret, "failed to register debugfs dev\n");
+	}
+
+	dev_set_drvdata(hsotg->ddev, hsotg);
+	ret = devm_pm_runtime_enable(hsotg->ddev);
+	if (ret)
+		return dev_err_probe(hsotg->ddev, ret, "Failed to enable pm runtime\n");
+
 	root = debugfs_create_dir(dev_name(hsotg->dev), usb_debug_root);
 	hsotg->debug_root = root;
 
@@ -795,12 +932,15 @@ int dwc2_debugfs_init(struct dwc2_hsotg
 	hsotg->regset->regs = dwc2_regs;
 	hsotg->regset->nregs = ARRAY_SIZE(dwc2_regs);
 	hsotg->regset->base = hsotg->regs;
+	hsotg->regset->dev = hsotg->ddev; /* Manage runtime PM for regdump */
 
 	debugfs_create_regset32("regdump", 0444, root, hsotg->regset);
 
 	return 0;
 err:
 	debugfs_remove_recursive(hsotg->debug_root);
+	device_unregister(hsotg->ddev);
+
 	return ret;
 }
 
@@ -808,4 +948,7 @@ void dwc2_debugfs_exit(struct dwc2_hsotg
 {
 	debugfs_remove_recursive(hsotg->debug_root);
 	hsotg->debug_root = NULL;
+	/* dwc2_debugfs_init() result may be ignored, so check if ddev has been registered */
+	if (hsotg->ddev)
+		device_unregister(hsotg->ddev);
 }
--- a/drivers/usb/dwc2/drd.c
+++ b/drivers/usb/dwc2/drd.c
@@ -210,6 +210,9 @@ static int dwc2_drd_role_sw_set(struct u
 		goto skip;
 	}
 
+	/* prevent runtime suspend while changing role */
+	pm_runtime_get(hsotg->dev);
+
 	if ((IS_ENABLED(CONFIG_USB_DWC2_PERIPHERAL) ||
 	     IS_ENABLED(CONFIG_USB_DWC2_DUAL_ROLE)) &&
 	     dwc2_is_device_mode(hsotg) &&
@@ -274,6 +277,8 @@ static int dwc2_drd_role_sw_set(struct u
 		spin_unlock_irqrestore(&hsotg->lock, flags);
 	}
 
+	pm_runtime_put(hsotg->dev);
+
 skip:
 	if (!hsotg->ll_hw_enabled && hsotg->clk)
 		clk_disable_unprepare(hsotg->clk);
@@ -283,14 +288,16 @@ skip:
 		role == USB_ROLE_HOST ? "A" : "B");
 
 	if (role == USB_ROLE_NONE && !hsotg->rpm_suspended) {
+		/* Enforce dwc2_conn_id_status_change has completed */
+		if (hsotg->wq_otg)
+			flush_workqueue(hsotg->wq_otg);
 		/*
 		 * With none role, the controller can be put in low power mode,
 		 * in case of a new event, e.g. usb-role-switch, role_sw_set
 		 * will be called, to resume.
 		 */
-		spin_lock_irqsave(&hsotg->lock, flags);
-		dwc2_gadget_enter_lp(hsotg);
-		spin_unlock_irqrestore(&hsotg->lock, flags);
+		hsotg->rpm_suspended = true;
+		pm_runtime_put_sync(hsotg->dev);
 	}
 
 	return 0;
--- a/drivers/usb/dwc2/gadget.c
+++ b/drivers/usb/dwc2/gadget.c
@@ -3279,7 +3279,10 @@ static void dwc2_gadget_enum_timeout(str
 	dev_dbg(hsotg->dev, "%s enumeration timeout\n",  __func__);
 
 	spin_lock_irqsave(&hsotg->lock, flags);
-	dwc2_gadget_enter_lp(hsotg);
+	if (!hsotg->rpm_suspended) {
+		pm_runtime_put(hsotg->dev);
+		hsotg->rpm_suspended = true;
+	}
 	spin_unlock_irqrestore(&hsotg->lock, flags);
 }
 
@@ -4787,7 +4790,10 @@ static int dwc2_hsotg_udc_stop(struct us
 	if (!IS_ERR_OR_NULL(hsotg->uphy))
 		otg_set_peripheral(hsotg->uphy->otg, NULL);
 
-	dwc2_gadget_enter_lp(hsotg);
+	if (!hsotg->rpm_suspended) {
+		hsotg->rpm_suspended = true;
+		pm_runtime_put_sync(hsotg->dev);
+	}
 
 	if (hsotg->dr_mode == USB_DR_MODE_PERIPHERAL)
 		dwc2_lowlevel_hw_disable(hsotg);
@@ -6018,12 +6024,15 @@ void dwc2_gadget_exit_clock_gating(struc
  */
 int dwc2_gadget_enter_lp(struct dwc2_hsotg *hsotg)
 {
+	unsigned long flags;
 	int ret;
 
+	spin_lock_irqsave(&hsotg->lock, flags);
 	switch (hsotg->params.power_down) {
 	case DWC2_POWER_DOWN_PARAM_HIBERNATION:
 		ret = dwc2_gadget_enter_hibernation(hsotg);
 		if (ret) {
+			spin_unlock_irqrestore(&hsotg->lock, flags);
 			dev_err(hsotg->dev, "enter hibernation failed %d\n", ret);
 			return ret;
 		}
@@ -6032,6 +6041,7 @@ int dwc2_gadget_enter_lp(struct dwc2_hso
 	case DWC2_POWER_DOWN_PARAM_PARTIAL:
 		ret = dwc2_gadget_enter_partial_power_down(hsotg);
 		if (ret) {
+			spin_unlock_irqrestore(&hsotg->lock, flags);
 			dev_err(hsotg->dev, "enter partial_power_down failed %d\n", ret);
 			return ret;
 		}
@@ -6045,6 +6055,7 @@ int dwc2_gadget_enter_lp(struct dwc2_hso
 		if (!hsotg->params.no_clock_gating)
 			dwc2_gadget_enter_clock_gating(hsotg);
 	}
+	spin_unlock_irqrestore(&hsotg->lock, flags);
 
 	/* Core hasn't been put into one of the power saving modes: simply return */
 	if (!(hsotg->bus_suspended || hsotg->in_ppd || hsotg->hibernated))
@@ -6054,10 +6065,5 @@ int dwc2_gadget_enter_lp(struct dwc2_hso
 	if (!IS_ERR_OR_NULL(hsotg->uphy))
 		usb_phy_set_suspend(hsotg->uphy, true);
 
-	if (!hsotg->rpm_suspended) {
-		pm_runtime_put(hsotg->dev);
-		hsotg->rpm_suspended = true;
-	}
-
 	return 0;
 }
--- a/drivers/usb/dwc2/hcd.c
+++ b/drivers/usb/dwc2/hcd.c
@@ -3318,7 +3318,8 @@ int dwc2_port_suspend(struct dwc2_hsotg
 
 	if (!hsotg->rpm_suspended) {
 		spin_unlock_irqrestore(&hsotg->lock, flags);
-		dwc2_host_enter_lp(hsotg);
+		hsotg->rpm_suspended = true;
+		pm_runtime_put_sync(hsotg->dev);
 		spin_lock_irqsave(&hsotg->lock, flags);
 	}
 
@@ -4359,7 +4360,8 @@ static int _dwc2_hcd_suspend(struct usb_
 
 	if (!hsotg->rpm_suspended) {
 		spin_unlock_irqrestore(&hsotg->lock, flags);
-		dwc2_host_enter_lp(hsotg);
+		hsotg->rpm_suspended = true;
+		pm_runtime_put_sync(hsotg->dev);
 		spin_lock_irqsave(&hsotg->lock, flags);
 
 		/* After entering suspend, hardware is not accessible */
@@ -5773,8 +5775,6 @@ static int dwc2_host_enter_partial_power
 	hsotg->lx_state = DWC2_L2;
 	hsotg->bus_suspended = true;
 
-	pm_runtime_put(hsotg->dev);
-	hsotg->rpm_suspended = true;
 	dev_dbg(hsotg->dev, "Entering host partial power down completed.\n");
 
 	return ret;
@@ -5919,9 +5919,6 @@ static void dwc2_host_enter_clock_gating
 
 	hsotg->bus_suspended = true;
 	hsotg->lx_state = DWC2_L2;
-
-	pm_runtime_put(hsotg->dev);
-	hsotg->rpm_suspended = true;
 }
 
 /**
--- a/drivers/usb/dwc2/platform.c
+++ b/drivers/usb/dwc2/platform.c
@@ -1078,8 +1078,65 @@ static int __maybe_unused dwc2_resume(st
 	return ret;
 }
 
+/*
+ * Only power saving mode entry is delegated to PM runtime suspend routine.
+ * This way, power saving mode is entered, only after all children have
+ * supended themselves.
+ *
+ * All the resume activities, remains in each child driver part, e.g.:
+ * - gadget resume or remote wakeup IRQ, calls directly the dwc2_gadget_exit_xxx
+ * - hcd port or bus resume routines, calls directly the dwc2_host_exit_xxx
+ * - debugfs runtime resume routine, calls directly...
+ *
+ * Moving resume activities into a runtime resume routine would have a side effect
+ * in gadget mode. A resume IRQ would be triggered, upon system-wide suspend entry:
+ * the (unused) root hub resumes to reconfigure the port without wakeup capability
+ * (as there's no device plugged), without a real need to update dwc2 configuration.
+ * But, this would abort a sequence with a suspended gadget capable of doing remote
+ * wakeup, e.g. bring it out of low power.
+ */
+static int dwc2_runtime_suspend(struct device *dev)
+{
+	struct dwc2_hsotg *hsotg = dev_get_drvdata(dev);
+	int ret;
+
+	dev_dbg(hsotg->dev, "%s lx_state %d\n", __func__, hsotg->lx_state);
+
+	if (dwc2_is_device_mode(hsotg))
+		ret = dwc2_gadget_enter_lp(hsotg);
+	else
+		ret = dwc2_host_enter_lp(hsotg);
+
+	if (ret)
+		return ret;
+
+	/* If the core hasn't been put into power saving modes, keep PM domain active */
+	if (!(hsotg->bus_suspended || hsotg->in_ppd || hsotg->hibernated))
+		return -EBUSY;
+
+	return 0;
+}
+
+static int dwc2_runtime_idle(struct device *dev)
+{
+	struct dwc2_hsotg *hsotg = dev_get_drvdata(dev);
+
+	/* RPM suspend not requested by the driver */
+	if (!hsotg->rpm_suspended)
+		return -EBUSY;
+
+	/* Connected device isn't suspended */
+	if (dwc2_is_device_mode(hsotg)) {
+		if (dwc2_is_device_connected(hsotg) && !(dwc2_readl(hsotg, DSTS) & DSTS_SUSPSTS))
+			return -EBUSY;
+	}
+
+	return 0;
+}
+
 static const struct dev_pm_ops dwc2_dev_pm_ops = {
 	SET_SYSTEM_SLEEP_PM_OPS(dwc2_suspend, dwc2_resume)
+	RUNTIME_PM_OPS(dwc2_runtime_suspend, NULL, dwc2_runtime_idle)
 };
 
 static struct platform_driver dwc2_platform_driver = {
